File: problems/2304_minimum_path_cost_in_a_grid/Solution.java
import java.util.*;

class Solution {
    public int minPathCost(int[][] grid, int[][] moveCost) {
        int m = grid.length, n = grid[0].length;
        int[] dp = grid[0].clone();
        for (int i = 1; i < m; i++) {
            int[] newDp = new int[n];
            Arrays.fill(newDp, Integer.MAX_VALUE);
            for (int j = 0; j < n; j++)
                for (int k = 0; k < n; k++)
                    newDp[j] = Math.min(newDp[j], dp[k] + moveCost[grid[i-1][k]][j] + grid[i][j]);
            dp = newDp;
        }
        int res = Integer.MAX_VALUE;
        for (int v : dp) res = Math.min(res, v);
        return res;
    }
}

File: problems/2304_minimum_path_cost_in_a_grid/solution.py
from typing import List


class Solution:
    def minPathCost(self, grid: List[List[int]], moveCost: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dp = [row[:] for row in [grid[0]]]
        dp = grid[0][:]
        for i in range(1, m):
            new_dp = [float('inf')] * n
            for j in range(n):
                for k in range(n):
                    new_dp[j] = min(new_dp[j], dp[k] + moveCost[grid[i-1][k]][j] + grid[i][j])
            dp = new_dp
        return min(dp)
